### PR TITLE
Port the perception scoring plugin

### PR DESCRIPTION
**Pull request by caguero**
Let's wait until #523 is merged so that we can use some of the bridge infrastructure there.

### How to test it?

1. Launch the simulation loading the new `perception_task` world:
```
ros2 launch vrx_gz competition.launch.py world:=perception_task
```

2. Verify that at `t=25`, two buoys (red and green) are spawned ~8m in front of the WAM-V. The buoys should stay there for another 5 seconds, and then, they should disappear from the field of view.

3. Verify that at `t=35`, one buoy (black) is spawned ~8m in front of the WAM-V. The buoy should stay there for another 5 seconds, and then, it should disappear from the field of view.

4. The simulation should stop at `t=40`.

perception_task

For testing the perception reports:

1. Create a directory for the client:
```
mkdir ~/perception_client & cd perception_client
```
2. Download these three files from [here](https://gist.github.com/caguero/2dd7ae577ab4d535e0c4d9f762e44163).
3. Create a build directory and compile:
```
mkdir build && cd build
cmake ..
make
```
4. Launch the simulation as the previous (1) step.
5. When you see the first round of objects, launch the client to report the red buoy:
```
. /opt/ros/humble/setup.bash
./perception_client
```
6. Wait until the simulation ends and look for the perception summary:
```
[ruby $(which gz) sim-1] [Dbg] [PerceptionScoringPlugin.cc:486] PerceptionScoring: 
[ruby $(which gz) sim-1] 	Name: red_0
[ruby $(which gz) sim-1] 	Type: mb_marker_buoy_red
[ruby $(which gz) sim-1] 	Time: 5
[ruby $(which gz) sim-1] 	Duration: 5
[ruby $(which gz) sim-1] 	Orig pose: -200 100 1 0 0 0
[ruby $(which gz) sim-1] 	Trial pose: 8 0 1 0 0 0
[ruby $(which gz) sim-1] 	Entity: 112
[ruby $(which gz) sim-1] 	Active: false
[ruby $(which gz) sim-1] 	Error: 0.0057877
[ruby $(which gz) sim-1] 
[ruby $(which gz) sim-1] [Dbg] [PerceptionScoringPlugin.cc:486] PerceptionScoring: 
[ruby $(which gz) sim-1] 	Name: green_0
[ruby $(which gz) sim-1] 	Type: mb_marker_buoy_green
[ruby $(which gz) sim-1] 	Time: 5
[ruby $(which gz) sim-1] 	Duration: 5
[ruby $(which gz) sim-1] 	Orig pose: -200 102 1 0 0 0
[ruby $(which gz) sim-1] 	Trial pose: 8 2 1 0 0 0
[ruby $(which gz) sim-1] 	Entity: 117
[ruby $(which gz) sim-1] 	Active: false
[ruby $(which gz) sim-1] 	Error: 10
[ruby $(which gz) sim-1] 
[ruby $(which gz) sim-1] [Dbg] [PerceptionScoringPlugin.cc:486] PerceptionScoring: 
[ruby $(which gz) sim-1] 	Name: black_0
[ruby $(which gz) sim-1] 	Type: mb_marker_buoy_black
[ruby $(which gz) sim-1] 	Time: 15
[ruby $(which gz) sim-1] 	Duration: 5
[ruby $(which gz) sim-1] 	Orig pose: -200 104 1 0 0 0
[ruby $(which gz) sim-1] 	Trial pose: 8 -2 1 0 0 0
[ruby $(which gz) sim-1] 	Entity: 122
[ruby $(which gz) sim-1] 	Active: false
[ruby $(which gz) sim-1] 	Error: 10
[ruby $(which gz) sim-1] 
[ruby $(which gz) sim-1] [Dbg] [PerceptionScoringPlugin.cc:490] Perception run score: 6.6686
```

Notice how the `mb_markler_buoy_red` buoy has an almost `0` error and the overall score is `6.66`. If you don't send the report you should get a `10`.

`Note`: It should be possible and way easier to send the report via `ros2 topic pub` but my autocomplete didn't really help me. I had a lot of escape characters and I couldn't fill the message properly.

~~**Note:** I'll update these instructions when #523 is merged to report the objects via ROS 2.~~

### CODE REVIEW

**Comment by iche033** [comment]
verified that the buoys appears and disappears correctly. 

I don't see the `/vrx/perception/landmark` ros2 topic though

**Comment by caguero** [comment]
> verified that the buoys appears and disappears correctly.
> 
> I don't see the `/vrx/perception/landmark` ros2 topic though

The ros2 topic should be working now.

File: vrx_gz/worlds/perception_task.sdf [hunk 1]
```diff
@@ -0,0 +1,606 @@
...
+    <plugin
+      filename="libPerceptionScoringPlugin.so"
+      name="vrx::PerceptionScoringPlugin">
+      <!-- Parameters for ScoringPlugin base class -->
+      <vehicle>wamv</vehicle>
+      <task_name>perception</task_name>
+      <initial_state_duration>10.0</initial_state_duration>
+      <ready_state_duration>10.0</ready_state_duration>
+      <running_state_duration>30</running_state_duration>
+      <!-- Parameters for PopulationPlugin -->
```
Comment (iche033) [inline vrx_gz/worlds/perception_task.sdf:577]: is there a `PopulationPlugin`?

File: vrx_gz/src/PerceptionScoringPlugin.cc [hunk 2]
```diff
@@ -0,0 +1,518 @@
...
+      return;
+    }
+
+    // In the ROS side, the object type is reported in the "frame_id" field
+    // of the header.
+    std::string typeReported;
+    for (auto i = 0u; i < _msg.header().data_size(); ++i)
+    {
+      if (_msg.header().data(i).key() == "frame_id")
+        typeReported = _msg.header().data(i).value(0);
```
Comment (iche033) [inline vrx_gz/src/PerceptionScoringPlugin.cc:365]: `break` as well?

**Review by iche033 — APPROVED** [review]
changes look good. Just very minor comments.

File: vrx_gz/worlds/perception_task.sdf (hunk 1 above)
Comment (caguero) [reply vrx_gz/worlds/perception_task.sdf]: Renamed in [950eb68](https://github.com/osrf/vrx/pull/525/commits/950eb68cb21b687e12d0b8146b8e74aa7cfe9dbd).

File: vrx_gz/src/PerceptionScoringPlugin.cc (hunk 2 above)
Comment (caguero) [reply vrx_gz/src/PerceptionScoringPlugin.cc:366]: Added in [950eb68](https://github.com/osrf/vrx/pull/525/commits/950eb68cb21b687e12d0b8146b8e74aa7cfe9dbd).

**Comment by iche033** [comment]
perception test works for me! We should probably set up CI and add this as an integration test later

**Pull request merged**